File: A-Star.RS/src/heap.rs
struct Heap {

    data: Box<[*mut Node]>,
    count: mut u32,

}

//

impl Heap {

    pub fn new(max_size: u32) -> Box<Heap> {
        box Heap {
            data: Box::new([*mut Node; max_size]),
            count: 0,
        }
    }


    pub fn Add(&self, node: &Node) {
        self.data[self.count] = node;
        node._heap_index = self.count;

        self.PercolateUp(self.count);

        self.count++;
    }


    fn PercolateUp(&self, index: u32) {
        let parent_index = (index - 1) / 2;

        let parent: *Node = &self.data[parent_index];
        let node: *Node = &self.data[index];

        if Node.Compare(node, parent) > 0 {
            self.Swap(index, parent_index);
            self.PercolateUp(parent_index);
        }
    }

    fn PercolateDown(&self, index: u32) {
        left_index: u32 = 2 * index + 1;
        right_index: u32 = 2 * index + 2;

        swap_index: mut u32 = 0;

        if left_index < self.count {
            swap_index = left_index;

            if right_index < self.count {
                if self.data[right_index].F() < self.data[left_index].F() {
                    swap_index = right_index;
                }
            }

            if Node.Compare(self.data[index], self.data[swap_index]) < 0 {
                self.Swap(index, swap_index);
                self.PercolateDown(swap_index);
            }
        }
    }

    fn Swap(&self, first: u32, second: u32) {
        let temp: *Node = &self.data[first];

        self.data[first] = &self.data[second];
        self.data[second] = &temp;
    }

}

File: main/src/Heap.cs
using System;
using System.Collections.Generic;

namespace AStar {

	internal class NodeHeap {

		Node[] data;
		int count;


		//----------------------------------------------------------------------------------------------------------------------------------<


		public NodeHeap(int maxSize) {
			data = new Node[maxSize];
			count = 0;
		}


		//----------------------------------------------------------------------------------------------------------------------------------<


		public int Count => count;
		public Node Peek(int i) => data[i];
		public bool Contains(Node node) => node == data[node.HeapIndex];
		public void Update(Node node) => PercolateUp(node.HeapIndex);
		public void Remove(Node node) => Remove(node.HeapIndex);


		//----------------------------------------------------------------------------------------------------------------------------------<


		public void Add(Node node) {
			data[count] = node;
			node.HeapIndex = count;

			PercolateUp(count);

			count++;
		}


		//----------------------------------------------------------------------------------------------------------------------------------<


		public Node Remove(int i) {
			Node node = data[i];
			node.HeapIndex = 0;

			count--;

			data[i] = data[count];
			data[i].HeapIndex = i;

			data[count] = null;
			
			PercolateDown(i);

			return node;
		}


		//----------------------------------------------------------------------------------------------------------------------------------<


		private void PercolateUp(int i) {
			int parentIndex = (i - 1) / 2;

			Node parent = data[parentIndex];
			Node node = data[i];

			if(node.Compare(parent) < 0) {
				Swap(i, parentIndex);
				PercolateUp(parentIndex);
			}
		}


		//----------------------------------------------------------------------------------------------------------------------------------<


		private void PercolateDown(int i) {
			int leftIndex = 2 * i + 1;
			int rightIndex = 2 * i + 2;

			int swapIndex;

			if(leftIndex < count) {
				swapIndex = leftIndex;

				if(rightIndex < count) {
					if(data[rightIndex].F < data[leftIndex].F)
						swapIndex = rightIndex;
				}

				if(data[i].Compare(data[swapIndex]) > 0) {
					Swap(i, swapIndex);
					PercolateDown(swapIndex);
				}
			}
		}


		//----------------------------------------------------------------------------------------------------------------------------------<


		private void Swap(int a, int b) {
			Node na = data[a];
			Node nb = data[b];

			data[a] = nb;
			data[b] = na;

			na.HeapIndex = b;
			nb.HeapIndex = a;
		}
	}

}
